multistreaming-server/rtmp-conf-generator.py

```diff
@@ -123,42 +141,64 @@ def generatePlatormPushURL(block_config):
 def createRTMPApplicationBlocks(block_name, block_config):
     app_block = ''
     primary_block_name = block_name
-    if 'transcode' in block_config:
+    is_icecast = block_config['platform'] == 'icecast'
+    audio_opts = ''
+    if 'transcode' in block_config or is_icecast:
         primary_block_name += '_transcoded'
         tc_conf = block_config['transcode']
         pixel_size = tc_conf['pixels'] if 'pixels' in tc_conf else '1280x720'
         video_bit_rate = tc_conf['videoBitRate'] if 'videoBitRate' in tc_conf else '4500k'
         key_frames = 30 * \
             tc_conf['videoKeyFrameSecs'] if 'videoKeyFrameSecs' in tc_conf else 60
-        if ('audioBitRate' in tc_conf) or ('audioSampleRate' in tc_conf):
+        if ('audioCodec' in tc_conf) or ('audioBitRate' in tc_conf) or ('audioSampleRate' in tc_conf):
+            audio_codec = tc_conf['audioCodec'] if 'audioCodec' in tc_conf else 'libfdk_aac'
             audio_bit_rate = tc_conf['audioBitRate'] if 'audioBitRate' in tc_conf else '160k'
             audio_sample_rate = str(tc_conf['audioSampleRate']) if 'audioSampleRate' in tc_conf else '48000'
             audio_opts = RTMP_TRANSCODE_AUDIO_OPTS_CUSTOM.replace(
+                    '%%AUDIO_CODEC%%', audio_codec
+                ).replace(
                     '%%AUDIO_BIT_RATE%%', audio_bit_rate
                 ).replace(
                     '%%AUDIO_SAMPLE_RATE%%', audio_sample_rate
-                    )
+                )
         else:
             audio_opts = RTMP_TRANSCODE_AUDIO_OPTS_COPY
 
-        app_block += RTMP_TRANSCODE_BLOCK.replace(
+        if not is_icecast:
+            app_block += RTMP_TRANSCODE_BLOCK.replace(
+                    '%%BLOCK_NAME%%', block_name
+                ).replace(
+                    '%%DEST_BLOCK_NAME%%', primary_block_name
+                ).replace(
+                    '%%PIXEL_SIZE%%', pixel_size
+                ).replace(
+                    '%%VIDEO_BIT_RATE%%', video_bit_rate
+                ).replace(
+                    '%%KFS%%', str(key_frames)
+                ).replace(
+                    '%%AUDIO_OPTS%%', audio_opts
+                )
+
+    if is_icecast:
```

Comment:
You use `is_icecast` multiple times in this function to separate mostly similar code. That points to maybe we should consider break this function up some to better encapsulate responsibilities. Maybe factor out the audio transcoding configuration? Maybe having a function for the `app_block` generation that gets passed the block template it need to adjust? Remember, calling `replace()` on a string for a token the string does not have has no net effect on the string. Yeah, maybe it waste a few CPU cycles, but this isn't a compute intensive python program anyway.

Reply:
Yeah, I thought about that, and did it somewhat that way in the previous pull request. You wanted to keep the logic more together, so I decided to solve it this way. I will break it up into more functions again.

Reply:
I think maybe the problem last time is you had an icecast specific function, rather than a general function that can be used for all cases. in this implementation, there are similarities betwen the `is_icecast` and `!is_icecast` cost. What I am ask is can those similarities be reduced to a single implementation?